lib/timers.js

```diff
@@ -211,9 +211,9 @@ function listOnTimeout() {
   debug('%d list empty', msecs);
   assert(L.isEmpty(list));
   this.close();
-  if (list._unrefed === true) {
+  if (list._unrefed === true && list === unrefedLists[msecs]) {
```

Comment:
Do you mind adding a block comment that describes why these checks are needed? Maybe something very similar to [the block comment included with the newly added test](https://github.com/nodejs/node/pull/7827/files#diff-d95655ffcc984cf8c55e8bf0388673f1R6) would be a good fit.

The goal would be to make it clear to anyone reading the code (including those not familiar with it) why these extra checks are needed.